references/common/metrics.go

```diff
@@ -0,0 +1,193 @@
+	}
+	return m, nil
+}
+
+// GetPodStorage get pod storage
+func GetPodStorage(client client.Client, pod v1.Pod) (storages []v1.PersistentVolumeClaim) {
+	for _, v := range pod.Spec.Volumes {
+		if v.PersistentVolumeClaim != nil {
+			storage := v1.PersistentVolumeClaim{}
+			err := client.Get(context.Background(), types.NamespacedName{Name: v.PersistentVolumeClaim.ClaimName, Namespace: pod.Namespace}, &storage)
+			if err != nil {
+				continue
+			}
+			storages = append(storages, storage)
+		}
+	}
+	return
+}
+
+// ListApplicationResource list application resource
+func ListApplicationResource(c client.Client, name, namespace string) ([]query.Resource, error) {
+	opt := query.Option{
+		Name:      name,
+		Namespace: namespace,
+		Filter:    query.FilterOption{},
+	}
+
+	collector := query.NewAppCollector(c, opt)
+	appResList, err := collector.CollectResourceFromApp(context.Background())
+	if err != nil {
+		return []query.Resource{}, err
+	}
+	return appResList, err
+}
+
+// ListApplicationPods list application pods
+func ListApplicationPods(c client.Client, app *appv1beta1.Application, components []string) []v1.Pod {
+	pods := make([]v1.Pod, 0)
+	opt := query.Option{
+		Name:      app.Name,
+		Namespace: app.Namespace,
+		Filter: query.FilterOption{
+			Components: components,
+			APIVersion: "v1",
+			Kind:       "Pod",
+		},
+		WithTree: true,
+	}
+
+	objects, err := CollectApplicationResource(context.Background(), c, opt)
+	if err != nil {
+		return pods
+	}
+	for _, object := range objects {
+		pod := v1.Pod{}
+		err = runtime.DefaultUnstructuredConverter.FromUnstructured(object.UnstructuredContent(), &pod)
+		if err != nil {
+			continue
+		}
+		pods = append(pods, pod)
+	}
+	return pods
+}
+
+// LoadApplicationMetrics load application metrics
+func (appMetrics *ApplicationMetricsStatus) LoadApplicationMetrics(c client.Client, conf *rest.Config, pods []v1.Pod) {
+	metricsClient := metricsclientset.NewForConfigOrDie(conf)
+	cpu, memory, stroage := uint64(0), uint64(0), uint64(0)
+
+	for _, pod := range pods {
+		podMetrics, err := GetPodMetrics(metricsClient, pod, false)
+		if err != nil {
+			continue
+		}
+
+		for _, containerMetrics := range podMetrics.Containers {
+			cpu += uint64(containerMetrics.Usage.Cpu().MilliValue())
+			memory += uint64(containerMetrics.Usage.Memory().Value() / (1024 * 1024))
+		}
+
+		storages := GetPodStorage(c, pod)
+		for _, s := range storages {
+			stroage += uint64(s.Status.Capacity.Storage().Value() / (1024 * 1024 * 1024))
+		}
+	}
+	appMetrics.CPU = cpu
+	appMetrics.Memory = memory
+	appMetrics.Storage = stroage
+}
+
+// LoadApplicationMetrics load application resource metrics
+func LoadApplicationMetrics(c client.Client, conf *rest.Config, app *appv1beta1.Application) (*ApplicationMetrics, error) {
+	appResList, err := ListApplicationResource(c, app.Name, app.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	components := make([]string, 0)
+	for _, resource := range appResList {
```

Comment:
Does `appResList` only have Workload? Do it has Service, ConfigMap or other resources?
Non-workload resource does not have pod.

Reply:
I think some questions can be answer at same time. Now we use `*v1beta1.ResourceTracker` object to track application's resource. So, if resource belong to an application, it can be tracked out.

Reply:
When we statistic the number of Managed Resource, I think Service, ConfigMap or other resources should be counted. Because we just statistic the son resource of an application, although it doesn't have pod